include/deal.II/lac/affine_constraints.templates.h

```diff
@@ -992,6 +992,37 @@ AffineConstraints<number>::close()
          ExcMessage("The constraints represented by this object have a cycle. "
                     "This is not allowed."));
 
+  // Now also compute which indices we need in distribute().
+  //
+  // This processor owns only part of the vector. one may think that
+  // every processor should be able to simply communicate those elements
+  // it owns and for which it knows that they act as sources to constrained
+  // DoFs to the owner of these DoFs. This would lead to a scheme where all
+  // we need to do is to add some local elements to (possibly non-local)
+  // ones and then call compress().
+  //
+  // Alas, this scheme does not work as evidenced by the disaster of bug
+  // #51, see http://code.google.com/p/dealii/issues/detail?id=51 and the
+  // reversion of one attempt that implements this in r29662. Rather, we
+  // need to get a vector that has all the *sources* or constraints we
+  // own locally, possibly as ghost vector elements, then read from them,
+  // and finally throw away the ghosted vector. Implement this in the
+  // following.
+  needed_elements_for_distribute = local_lines;
+
+  if (needed_elements_for_distribute != complete_index_set(local_lines.size()))
+    {
+      std::vector<types::global_dof_index> additional_elements;
```

Comment:
Not sure if this is a bottleneck, but is there any indication of a memory footprint that could be used to pre-reserve some memory? I know that we won't know which line in lines are "elements",  but maybe the size of local_lines could give us a pre-emptive indication of the memory footprint?

Reply:
I don't know how to estimate this. It's the number of DoFs against which locally relevant DoFs are constrained. I don't think there is a good way of estimating this. One could duplicate the loop and in a first go-around count how many elements we get, and in the second actually insert them, calling `.reserve(...)` in between. But the calls to `is_element()` are expensive enough that I don't think that's a win :-(